fields/src/fp_256.rs

```diff
@@ -376,6 +376,86 @@ impl<P: Fp256Parameters> PrimeField for Fp256<P> {
     type BigInteger = BigInteger;
     type Parameters = P;
 
+    #[inline]
+    fn decompose(&self) -> (Self, Self, bool, bool) {
+        const Q1: [u64; 4] = [9183663392111466540, 12968021215939883360, 3, 0];
+        const Q2: [u64; 4] = [13, 0, 0, 0];
+        let b1 = Self::from_repr(BigInteger([725501752471715840, 4981570305181876225, 0, 0])).unwrap();
+        let b2 = Self::from_repr(BigInteger([725501752471715841, 4981570305181876225, 0, 0])).unwrap();
+        let r128 = Self::from_repr(BigInteger([0xffffffffffffffff, 0xffffffffffffffff, 0, 0])).unwrap();
+        const HALF_R: [u64; 8] = [0, 0, 0, 0x8000000000000000, 0, 0, 0, 0];
```

Comment:
These constants are specific to BLS12-377, right? We should move these to `BLS12Parameters`, and pass these in to `decompose`

Reply:
I'll put them on the `ShortWeierstrassParameters` as it's not possible to access `BLS12Parameters` from `Fp256` or `Projective` as far as I can tell.

Reply:
Hm I would rather put a `mul` method in `ShortWeierstrassParameters`, so that we can specify it when instantiating `Bls12377G1Parameters`. See for example https://github.com/arkworks-rs/algebra/blob/402e7f9603fca7a68b86baf296b6feaf904939f5/ec/src/models/short_weierstrass/mod.rs#L82

Reply:
Hmm, so as in a `mul_by_b1`, `mul_by_q1` etc. sort of thing?

Reply:
I mean by a `mul_projective(a: Projective<Self>, b: Self::ScalarField) → Projective<Self>`. This way you can encapsulate the GLV portion in the curve-specific parameters